### PR TITLE
Add xgboost sampler

### PR DESCRIPTION
**Pull request by AldoGl**
## Proposed changes

- I added the XGBoost sampler used in "Agent-based model calibration using machine learning surrogates"
- I added a test for the new sampler
- I added the XGBoost sampler to the Jupyter notebook with the samplers overview

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../blob/master/CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: examples/overview_of_the_different_samplers.ipynb [hunk 1]
```diff
@@ -39,7 +52,7 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 2,
+   "execution_count": 108,
```
Comment (marcofavoritobi) [inline examples/overview_of_the_different_samplers.ipynb:55]: (very picky) Could you fully execute the notebook from scratch so to make the execution count reasonable?

**Review by marcofavoritobi — APPROVED** [review]
LGTM, this is a useful feature.

I wonder whether we could develop a generalized sampler based on machine learning surrogates (i.e. regressors). Something like:


```python

class Regressor(ABC):

    @abstractmethod
    def fit(self, X, y):
        raise NotImplementedError

    @abstractmethod
    def predict(x):
        raise NotImplementedError


class SurrogateSampler(BaseSampler):


def __init__(  # pylint: disable=too-many-arguments
        self,
        batch_size: int,
        regressor: Regressor,
        random_state: Optional[int] = None,
        max_deduplication_passes: int = 5,
        candidate_pool_size: Optional[int] = None):
        ...
        self.regressor = regressor



```

File: examples/overview_of_the_different_samplers.ipynb (hunk 1 above)
Comment (AldoGl) [reply examples/overview_of_the_different_samplers.ipynb:55]: > LGTM, this is a useful feature.
> 
> I wonder whether we could develop a generalized sampler based on machine learning surrogates (i.e. regressors). Something like:
> 
> ```python
> class Regressor(ABC):
> 
>     @abstractmethod
>     def fit(self, X, y):
>         raise NotImplementedError
> 
>     @abstractmethod
>     def predict(x):
>         raise NotImplementedError
> 
> 
> class SurrogateSampler(BaseSampler):
> 
> 
>     def __init__(  # pylint: disable=too-many-arguments
>         self,
>         batch_size: int,
>         regressor: Regressor,
>         random_state: Optional[int] = None,
>         max_deduplication_passes: int = 5,
>         candidate_pool_size: Optional[int] = None):
>         ...
>         self.regressor = regressor
> ```

Yes I completely agree, it's something that I also thought I would do, but I wanted to do this in a different PR

**Comment by AldoGl** [comment]
Yes I completely agree, it's something that I also thought I would do, but I wanted to do this in a different PR

**Pull request merged**